mm2src/coins/coin_balance.rs

```diff
@@ -0,0 +1,452 @@
+    pub derivation_path: RpcDerivationPath,
+    pub new_addresses: Vec<HDAddressBalance>,
+}
+
+#[async_trait]
+pub trait EnableCoinBalanceOps {
+    async fn enable_coin_balance(&self) -> BalanceResult<EnableCoinBalance>;
+}
+
+#[async_trait]
+impl<Coin, Address, HDWallet, HDAccount, HDAddressChecker> EnableCoinBalanceOps for Coin
+where
+    Coin: AddressBalanceOps<Address = Address>
+        + CoinWithDerivationMethod<Address = Address, HDWallet = HDWallet>
+        + HDWalletBalanceOps<HDWallet = HDWallet, HDAccount = HDAccount, HDAddressChecker = HDAddressChecker>
+        + Sync,
+    Address: fmt::Display + Sync,
+    HDWallet: Sync,
+    HDAddressChecker: HDAddressBalanceChecker,
+{
+    async fn enable_coin_balance(&self) -> BalanceResult<EnableCoinBalance> {
+        match self.derivation_method() {
+            DerivationMethod::Iguana(address) => self.address_balance(address).await.map(|balance| {
+                EnableCoinBalance::Iguana(IguanaWalletBalance {
+                    address: address.to_string(),
+                    balance,
+                })
+            }),
+            DerivationMethod::HDWallet(hd_wallet) => self
+                .enable_hd_wallet_balance(hd_wallet)
+                .await
+                .map(EnableCoinBalance::HD),
+        }
+    }
+}
+
+#[async_trait]
+pub trait HDWalletBalanceOps: AddressBalanceOps {
+    type HDWallet;
+    type HDAccount;
+    type HDAddressChecker: HDAddressBalanceChecker<Address = Self::Address>;
+
+    async fn produce_hd_address_checker(&self) -> BalanceResult<Self::HDAddressChecker>;
+
+    /// Checks for addresses of every known account by using [`HDWalletBalanceOps::check_hd_account_balance`].
+    /// This method is used on coin initialization to index working addresses and to return the wallet balance to the user.
+    async fn enable_hd_wallet_balance(&self, hd_wallet: &Self::HDWallet) -> BalanceResult<HDWalletBalance>;
+
+    /// Checks for the new account addresses using the given `address_checker`.
+    /// Returns balances of new addresses.
+    async fn check_hd_account_balance(
+        &self,
+        hd_account: &mut Self::HDAccount,
+        address_checker: &Self::HDAddressChecker,
+        gap_limit: u32,
+    ) -> BalanceResult<Vec<HDAddressBalance>>;
+
+    /// Requests balance of the given `address`.
+    /// This function is expected to be more efficient than ['HDWalletBalanceOps::check_address_balance'] in most cases
+    /// since many of RPC clients allow us to request the address balance without the history.
+    async fn known_address_balance(&self, address: &Self::Address) -> BalanceResult<CoinBalance> {
+        self.address_balance(address).await
+    }
+
+    /// Checks if the address has been used by the user by checking if the transaction history of the given `address` is not empty.
+    /// Please note the function can return zero balance even if the address has been used before.
+    async fn check_address_balance(
```

Comment:
`check_address_was_used` seems to reflect the actual meaning.
Or even `is_address_used` that I noticed below.

Reply:
Done